internal/fpath/path.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package fpath
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// FPath creates an OS independently Path Handling Structure
+type FPath struct {
+	local  bool   //set if file is local
+	scheme string //url scheme
+	bucket string //set, when remote scheme
+	path   string //filepath OR remote path (without bucket)
+}
+
+// New creates Struct from handed over URL
+func New(url string) (p FPath, err error) {
+
+	// Check for Schema
+	sepstring := strings.SplitN(url, "://", 2)
+
+	switch len(sepstring) {
+	case 2: // Has Schema
+		p.scheme = sepstring[0]
+		if p.scheme == "sj" {
+			split := strings.SplitN(filepath.Clean(strings.TrimLeft(sepstring[1], "/")), `\`, 2)
+			if len(split) == 2 {
+				p.bucket = split[0]
+				p.path = filepath.ToSlash(split[1])
+			} else {
+				p.bucket = split[0]
+			}
+		} else {
+			p.path = sepstring[1]
+		}
+	case 1: // No Schema
+		p.local = true
+		p.path = sepstring[0]
+	default: // Everything else is misformatted
+		return FPath{}, fmt.Errorf("misformatted URL: %s", url)
+	}
+
+	// Check for Windows Special Handling Prefix
+	cprefix, _ := regexp.Compile(`^\\\\\?\\(UNC\\)?`)
+
+	// when Prefix present, omit further changes to the path
+	if prefix := cprefix.FindString(p.path); prefix != "" {
+
+		p.scheme = prefix
+		p.path = strings.Replace(p.path, prefix, "", 1) //Strip Prefix
+
+	} else {
+		// when file is local, ensure path absolute
+		if p.IsLocal() && !filepath.IsAbs(p.path) {
+
+			fullpath, err := filepath.Abs(p.path)
+
+			if err == nil {
+				p.path = fullpath
+			} else {
+				return FPath{}, fmt.Errorf("unable to create absolute path for: %s", p.path)
+			}
+		}
+	}
+	return p, nil
+}
+
+// Join is merging segment to the path
+func (p *FPath) Join(segment string) {
```

Comment:
Don't use a receiver here: `func (p FPath) Join(segment string) FPath`. With paths it's common to do things like:

```
a := dir.Join("a.txt")
b := dir.Join("b.txt")
```

Reply:
@egonelbre Which one do you like more?
```destObj.Join(srcObj.Base())
destObj := destObj.Join(scrObj.Base())```

Reply:
The second.

Reply:
Ok will change accordingly.
Was talking about that with @kaloyan-raev in a meeting and we both werent quite sure.